Extension/src/LanguageServer/configurations.ts

```diff
@@ -737,33 +738,58 @@ export class CppProperties {
 
     private resolveAndSplit(paths: string[] | undefined, defaultValue: string[] | undefined, env: Environment, glob = false): string[] {
         const result: string[] = [];
-        if (paths) {
-            paths = this.resolveDefaults(paths, defaultValue);
+        if (paths === undefined) {
+            return result;
+        }
+        paths = this.resolveDefaults(paths, defaultValue);
+        if (!glob) {
             paths.forEach(entry => {
                 const entries: string[] = util.resolveVariables(entry, env).split(util.envDelimiter).map(e => this.resolvePath(e, false)).filter(e => e);
                 result.push(...entries);
             });
-        }
-        if (!glob) {
             return result;
         }
 
-        const globResult: string[] = [];
-        for (let res of result) {
-            // fastGlob will expand the ending double wildcard. temporary strip them before expanding
-            const recursive: boolean = res.endsWith('**');
-            if (recursive) {
-                res = res.slice(0, res.length - 2);
+        for (let res of paths) {
+            let counter: number = 0;
+            let slashFound: boolean = false;
+            const lastIndex: number = res.length - 1;
+            // Detect and glob all wildcard variations by looking at last character in the path first.
+            for (let i: number = lastIndex; i >= 0; i--) {
+                if (res[i] === '*') {
+                    counter++;
+                } else if (res[i] === '/' || (res[i] === '\\' && isWindows)) {
+                    counter++;
+                    slashFound = true;
+                    break;
+                } else {
+                    break;
+                }
+            }
+            let suffix: string = '';
+            if (slashFound) {
+                suffix = res.slice(res.length - counter);
+                res = res.slice(0, res.length - counter);
+            }
+            let normalized: string = '';
+            let cwd: string = '';
+            if (isWindows) {
+                normalized = res.replace(/\\/g, '/');
+                cwd = this.rootUri?.fsPath?.replace(/\\/g, '/') || '';
             }
-            // fastGlob can't deal with backslash-separated path => remove them
-            const normalized: string = res.replace(/\\/g, '/');
-            const cwd: string = this.rootUri?.fsPath?.replace(/\\/g, '/') || '';
             // fastGlob silently strip non-found paths. limit that behavior to dynamic paths only
             const matches: string[] = fastGlob.isDynamicPattern(normalized) ?
-                fastGlob.sync(normalized, { onlyDirectories: true, cwd}) : [res];
-            globResult.push(...matches.map(s => recursive ? s + '**' : s));
+                fastGlob.sync(normalized, { onlyDirectories: true, cwd }) : [res];
+
+            result.push(...matches.map(s => s + suffix));
+            result.forEach(entry => {
```

Comment:
This doesn't appear correct. It's returning after 1 iteration of the for loop?

Reply:
I think because you're changing the value of `result` inside the callback.  You probably want to use `reduce` instead and push into the accumulation array (even though the name is counter-intuitive).

Reply:
But don't we need to resolve variables before the glob?

Reply:
We need to resolve the variables before the glob but we cannot normalize the paths before the globbing. I though the resolving the variables and normalizing the paths went hand in hand but it looks like I can possibly separate the two.